src/main/scala/org/broadinstitute/dsde/agora/server/model/AgoraEntityBsonSupport.scala

```diff
@@ -0,0 +1,57 @@
+package org.broadinstitute.dsde.agora.server.model
+
+import com.mongodb.DBObject
+import org.bson.types.ObjectId
+import org.joda.time.DateTime
+import org.joda.time.format.ISODateTimeFormat
+
+
+object AgoraEntityBsonSupport {
+
+  def read(obj: DBObject): AgoraEntity =
+    AgoraEntity(
+      namespace       = stringOrNone(obj, "namespace"),
+      name            = stringOrNone(obj, "name"),
+      snapshotId      = intOrNone(obj, "snapshotId"),
+      snapshotComment = stringOrNone(obj, "snapshotComment"),
+      synopsis        = stringOrNone(obj, "synopsis"),
+      documentation   = stringOrNone(obj, "documentation"),
+      owner           = stringOrNone(obj, "owner"), // Not in DB but checked by tests. Just rolling with this for now.
+      createDate      = dateOrNone(obj, "createDate"),
+      payload         = stringOrNone(obj, "payload"),
+      payloadObject   = None, // Not in DB
+      url             = None, // Not in DB
+      entityType      = entityType(obj),
+      id              = Option(new ObjectId(obj.get("_id").toString)),
+      methodId        = idOrNone(obj, "methodId"), // Configs only
+      method          = None, // Not in DB
+      managers        = Seq.empty, // Not in DB
+      public          = None, // Not in DB
+    )
+
+  private def idOrNone(obj: DBObject, key: String): Option[ObjectId] = {
+    if (obj.containsField(key)) Option(new ObjectId(obj.get(key).toString)) else None
+  }
+
+  private def stringOrNone(obj: DBObject, key: String): Option[String] = {
+    if (obj.containsField(key)) Option(obj.get(key).asInstanceOf[String]) else None
+  }
+
+  private def intOrNone(obj: DBObject, key: String): Option[Int] = {
+    if (obj.containsField(key)) Option(obj.get(key).asInstanceOf[Int]) else None
+  }
+
+  private def dateOrNone(obj: DBObject, key: String): Option[DateTime] = {
+    if (obj.containsField(key)) Option(DateTime.parse(obj.get(key).asInstanceOf[String], ISODateTimeFormat.dateTimeNoMillis())) else None
+  }
+
+  private def entityType(obj: DBObject): Option[AgoraEntityType.EntityType] = {
+    if (obj.containsField("entityType")) {
+      val entityString = Option(obj.get("entityType").asInstanceOf[String])
+      entityString.map(AgoraEntityType.withName)
+    } else {
+      None
+    }
+  }
```

Comment:
_Give yourself to the ~Dark Side~ Java `null`._ 😈

You can save on looking up the values twice by removing the calls to `containsField`.

Also, the `idOrNone`, `stringOrNone`, and `intOrNone` can be combined using generics into something like `getOrNone`.

```scala
private def getOrNone[A](obj: DBObject, key: String): Option[A] = {
  Option(obj.get(key).asInstanceOf[A])
}

private def dateOrNone(obj: DBObject, key: String): Option[DateTime] = {
  val dateString = getOrNone[String](obj, key)
  dateString.map(DateTime.parse(_, ISODateTimeFormat.dateTimeNoMillis()))
}

private def entityType(obj: DBObject): Option[AgoraEntityType.EntityType] = {
  val entityString = getOrNone[String](obj, "entityType")
  entityString.map(AgoraEntityType.withName)
}
```

Reply:
if `obj.get(key)` returns null, won't `obj.get(key).asInstanceOf[A]` throw NPE?

Reply:
No NPE, but good instincts anyway. You did find another bug fixed by your `Option(obj.get(key)).map(_.asInstanceOf[A])` with value types:

https://scastie.scala-lang.org/kshakir/Y6ktNXzvTs2CfsHbB28oeg/4